internal/vm/vsock.go

```diff
@@ -144,23 +228,93 @@ func VSockAcceptConnector(port uint32) IOConnector {
 	}
 }
 
-func isTemporaryNetErr(err error) bool {
-	terr, ok := err.(interface {
-		Temporary() bool
-	})
+func vsockConnectMsg(port uint32) string {
+	// The message a host-side connection must write after connecting to a firecracker
+	// vsock unix socket in order to establish a connection with a guest-side listener
+	// at the provided port number. This is specified in Firecracker documentation:
+	// https://github.com/firecracker-microvm/firecracker/blob/master/docs/vsock.md#host-initiated-connections
+	return fmt.Sprintf("CONNECT %d\n", port)
+}
 
-	return err != nil && ok && terr.Temporary()
+func vsockAckMsg(port uint32) string {
+	// The message a guest-side connection will write after accepting a connection from
+	// a host dial. This is not part of the official Firecracker vsock spec, but is
+	// recommended in order to allow the host to verify connections were established
+	// successfully: https://github.com/firecracker-microvm/firecracker/issues/1272#issuecomment-533004066
+	return fmt.Sprintf("IMALIVE %d\n", port)
+}
+
+// tryConnRead will try to do a read from the provided conn, returning an error if
+// the bytes read does not match what was provided or if the read does not complete
+// within the provided timeout. It will reset socket deadlines to none after returning.
+func tryConnRead(conn net.Conn, expectedRead string, timeout time.Duration) error {
+	conn.SetDeadline(time.Now().Add(timeout))
+	defer conn.SetDeadline(time.Time{})
+
+	actualRead := make([]byte, len(expectedRead))
+	_, err := conn.Read(actualRead)
+	if err != nil {
+		return err
+	}
+
+	if expectedRead != string(actualRead) {
+		return errors.Errorf("expected to read %q, but instead read %q",
+			expectedRead, string(actualRead))
+	}
+
+	return nil
+}
+
+// tryConnWrite will try to do a write to the provided conn, returning an error if
+// the write fails or does not complete within the provided timeout. It will reset
+// socket deadlines to none after returning.
+func tryConnWrite(conn net.Conn, expectedWrite string, timeout time.Duration) error {
+	conn.SetDeadline(time.Now().Add(timeout))
+	defer conn.SetDeadline(time.Time{})
+
+	bytesWritten, err := conn.Write([]byte(expectedWrite))
+	if err != nil {
+		return err
+	}
+	if bytesWritten != len(expectedWrite) {
```

Comment:
As discussed offline, since this isn't used as the main read/write implementation but is rather used only for the CONNECT and ack messages, I think that's fine.